pkg/controller/history/controller_history.go

```diff
@@ -0,0 +1,490 @@
+	"fmt"
+	"hash/fnv"
+	"sort"
+	"strconv"
+
+	apps "k8s.io/kubernetes/pkg/apis/apps/v1beta1"
+	"k8s.io/kubernetes/pkg/client/clientset_generated/clientset"
+	appsinformers "k8s.io/kubernetes/pkg/client/informers/informers_generated/externalversions/apps/v1beta1"
+	appslisters "k8s.io/kubernetes/pkg/client/listers/apps/v1beta1"
+	"k8s.io/kubernetes/pkg/controller"
+	hashutil "k8s.io/kubernetes/pkg/util/hash"
+
+	apiequality "k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/kubernetes/pkg/client/retry"
+)
+
+// ControllerRevisionHashLabel is the label used to indicate the hash value of a ControllerRevision's Data.
+const ControllerRevisionHashLabel = "controller.kubernetes.io/hash"
+
+// ControllerRevisionName returns the Name for a ControllerRevision in the form prefix-hash. If the length
+// of prefix is greater than 223 bytes, it is truncated to allow for a name that is no larger than 253 bytes.
+func ControllerRevisionName(prefix string, hash uint32) string {
+	if len(prefix) > 223 {
+		prefix = prefix[:223]
+	}
+	return fmt.Sprintf("%s-%d", prefix, hash)
+}
+
+// NewControllerRevision returns the a ControllerRevision with a ControllerRef pointing parent and indicating that
+// parent is of parentKind. The ControllerRevision has labels matching selector, contains Data equal to data, and
+// has a Revision equal to revision. If the returned error is nil, the returned ControllerRevision is valid. If the
+// returned error is not nil, the returned ControllerRevision is invalid for use.
+func NewControllerRevision(parent metav1.Object,
+	parentKind schema.GroupVersionKind,
+	selector labels.Selector,
+	data runtime.RawExtension,
+	revision int64) (*apps.ControllerRevision, error) {
+	labelMap, err := labels.ConvertSelectorToLabelsMap(selector.String())
+	if err != nil {
+		return nil, err
+	}
+	blockOwnerDeletion := true
+	isController := true
+	cr := &apps.ControllerRevision{
+		ObjectMeta: metav1.ObjectMeta{
+			Labels: labelMap,
+			OwnerReferences: []metav1.OwnerReference{
+				{
+					APIVersion:         parentKind.GroupVersion().String(),
+					Kind:               parentKind.Kind,
+					Name:               parent.GetName(),
+					UID:                parent.GetUID(),
+					BlockOwnerDeletion: &blockOwnerDeletion,
+					Controller:         &isController,
+				},
+			},
+		},
+		Data:     data,
+		Revision: revision,
+	}
+	hash := HashControllerRevision(cr, nil)
+	cr.Name = ControllerRevisionName(parent.GetName(), hash)
+	cr.Labels[ControllerRevisionHashLabel] = strconv.FormatInt(int64(hash), 10)
+	return cr, nil
+}
+
+// HashControllerRevision hashes the contents of revision's Data using FNV hashing. If probe is not nil, the byte value
```

Comment:
Nit: comment says "the byte value" is added to the hash, but it's the string representation of `*probe` that gets written.

Do you want to use `encoding/binary` and do `binary.Write(hf, binary.LittleEndian, *probe)` instead?